src/pointer/ptr.rs

```diff
@@ -1059,6 +1069,44 @@ mod _casts {
 mod _project {
     use super::*;
 
+    impl<'a, T, I> Ptr<'a, T, I>
+    where
+        T: 'a + ?Sized,
+        I: Invariants,
```

Comment:
Note that this can only ever be sound for invariants that structurally propagate to their fields. This doesn't apply for the `Pin` invariant.